text/0098-store-multiple-replay-segments-in-a-single-blob.md

```diff
@@ -0,0 +1,282 @@
+- Start Date: 2023-05-24
+- RFC Type: decision
+- RFC PR: https://github.com/getsentry/rfcs/pull/98
+- RFC Status: draft
+
+# Summary
+
+Recording data is sent in segments. Each segment is written to its own file. Writing files is the most expensive component of our Google Cloud Storage usage. It is also the most expensive component, in terms of time, in our processing pipeline. By merging many segment files together into a single file we can minimize our costs and maximize our processing pipeline's throughput.
+
+# Motivation
+
+1. Minimize costs.
+2. Increase write throughput.
+3. Enable new features in a cost-effective manner.
+
+# Background
+
+This document was originally written to respond to a percieved problem in the Session Replays recording consumer. However, upon exploring these ideas more it was determined that this could be more generally applied to the organization as a whole. For that reason I've made many of the names generic but have also retained many references to Session Replay.
+
+# Supporting Data
+
+Google Cloud Storage lists the costs for writing and storing data as two separate categories. Writing a file costs $0.005 per 1000 files. Storing that file costs $0.02 per gigabyte. For the average Session Replay file (with a retention period of 90 days) this works out to: $0.000000012 for storage and $0.00000005 for the write.
+
+In practical terms, this means 75% of our spend is allocated to writing new files.
+
+# High Level Overview
+
+1. We will store multiple "parts" per file.
+   - A "part" is a distinct blob of binary data.
+   - It exists as a subset of bytes within a larger set of bytes (referred to as a "file").
+   - A "part" could refer to a replay segment or to a sourcemap or anything that requires storage in a blob storage service.
+2. Each "part" within a file will be encrypted.
```

Comment:
Since encryption is used here entirely to make deletion of individual parts quicker, could you please expand on:
- whether we are confident that this is a use case worth optimizing for (deleting individual part quickly) 
- how would the process look like if we just accepted to rewrite files entirely in order to remove the part.

Reply:
> whether we are confident that this is a use case worth optimizing for (deleting individual part quickly)

It is necessary to support GDPR, project, and point deletes.  I consider it a high priority.  The alternative is rotating the file with the offending parts removed.

> how would the process look like if we just accepted to rewrite files entirely in order to remove the part.

The operations is no particular order:

1. Download the file
2. Get the active byte ranges from the database.
3. Remove all byte ranges from the file not found in the set of returned byte ranges.
4. Delete all references to the file in the database.
5. Upload the new file.
6. Insert new offset rows.
7. Delete the old file.

Repeat for every delete operation.  Deletes must be single-threaded per file to prevent concurrent access.  You can use kafka and partition on filename.